src/include/common/coroutine_defs.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <utility>
+
+#include <boost/coroutine2/all.hpp>
+
+#include "common/object_pool.h"
+
+namespace terrier::common {
+
+using pull_type = boost::coroutines2::coroutine<void>::pull_type;
+using push_type = boost::coroutines2::coroutine<void>::push_type;
+
+/**
+ * A reusable stack frame context which contains a workload function to be set.
+ *
+ * This context controls task switching with the execution pool, and may yield out of a running task on a call to
+ * YieldToPool() and temporarily stop on a task on a call to YieldToFunc() or complete the workload.
+ */
+class PoolContext {
+ public:
+  /**
+   * Sets the function associated with the current execution context
+   * @param f input function that will serve as workload for the pool context
+   */
+  void SetFunction(const std::function<void(PoolContext *)> &f) {
+    TERRIER_ASSERT(func_ == nullptr, "function should be null");
+    this->func_ = f;
+  }
+
+  /**
+   * Call will pause the workload and yield back to the execution pool.
+   */
+  void YieldToPool() {
+    TERRIER_ASSERT(sink_ != nullptr, "must have initialized sink_ before yielding to it");
+    (*sink_)();
+  }
+
+  /**
+   * @warning ONLY CALL IF YOU KNOW WHAT YOU'RE DOING this executes the function inside the context until it yields back
+   *
+   * Call will initialize or continue the resumeable workload from the last yield
+   * @return func_finished_ will signify whether the current function has finished its execution path
+   */
+  bool YieldToFunc() {
+    TERRIER_ASSERT(func_ != nullptr, "must have called SetFunction before yielding to function");
```

Comment:
I feel like the assertion can be combined with the next line, since both func_ and sink_ should be initialized